portiaui/app/utils/selectors.js

```diff
@@ -879,6 +879,71 @@ export function makeItemsFromGroups(groups) {
     return items;
 }
 
+function createSelectorGenerators(structure, selectorMatcher, accumulator) {
+    const generators = [];
+
+    for (let element of structure) {
+        const {annotation, children} = element;
+        let selectorGenerator;
+        if (children) {
+            selectorGenerator = ContainerSelectorGenerator.create({});
+            selectorGenerator.addChildren(
+                createSelectorGenerators(children, selectorMatcher, accumulator));
+
+        } else {
+            selectorGenerator = AnnotationSelectorGenerator.create({
+                selectorMatcher,
+                annotation
+            });
+        }
+        generators.push(selectorGenerator);
+        accumulator.push([annotation, selectorGenerator]);
+    }
+
+    return generators;
+}
+
+export function updateStructureSelectors(structure, selectorMatcher) {
+    const accumulator = [];
+    createSelectorGenerators(structure, selectorMatcher, accumulator);
+    for (let [annotation, selectorGenerator] of accumulator) {
+        const selector = selectorGenerator.get('selector');
+        if (selectorGenerator instanceof AnnotationSelectorGenerator) {
```

Comment:
the selector generator by design only operates on a structure format and doesn't need to know about the base annotation types that created the structure (it could be annotations and items, but in the future it may be link annotations, actions etc.) so it's the responsibility of the api consumer that created the structure to know how to apply the generated selectors back onto the annotations the structure was created from. having sait that, this clear that this belongs in annotation-structure.js